src/util/terminology.cpp

```diff
@@ -0,0 +1,246 @@
+    for(auto it = begin; it != to_resolve.end(); it++) {
+        const std::string &name = *it;
+        ptask_t task = boost::make_shared<task_t>(
+                boost::bind(resolveSingleNameInternal, context, session_ptr, name, terminology, key, match_type, first_hit, on_not_resolvable));
+        boost::shared_future<string_pair> future(task->get_future());
+        pending_results.push_back(future);
+        io_service.post(boost::bind(&task_t::operator(), task));
+    }
+
+    //get the resolved pairs from the futures, insert into map
+    for(auto &future : pending_results){
+        resolved_pairs.insert(future.get());
+    }
+
+    //insert values from resolved pairs into names_out
+    for(auto &name : names_in){
+        names_out.push_back(resolved_pairs[name]);
+    }
+    
+}
+
+/**
+ *
+ * @return a pair of strings, first string is the string to be resolved and the second is the resolved string.
+ */
+Terminology::string_pair Terminology::resolveSingleNameInternal(Poco::Net::Context::Ptr &context,
+                                                   Poco::Net::Session::Ptr &session_ptr,
+                                                   const std::string &name,
+                                                   const std::string &terminology,
+                                                   const std::string &key,
+                                                   const std::string &match_type,
+                                                   const bool first_hit,
+                                                   const HandleNotResolvable &on_not_resolvable)
+{
+    Poco::URI uri("https://terminologies.gfbio.org/api/terminologies/search");
+    uri.addQueryParameter("query", name);
+    uri.addQueryParameter("terminologies", terminology);
+    uri.addQueryParameter("match_type", match_type);
+    if(first_hit)
+        uri.addQueryParameter("first_hit", "true");
+
+    Poco::Net::HTTPSClientSession session(uri.getHost(), uri.getPort(), context, session_ptr);
+    Poco::Net::HTTPRequest request(Poco::Net::HTTPRequest::HTTP_GET, uri.getPathAndQuery(), Poco::Net::HTTPRequest::HTTP_1_1);
+    Poco::Net::HTTPResponse response;
+
+    session.sendRequest(request);
+    std::istream& respStream = session.receiveResponse(response);
+
+    //read response stream into json object
+    Json::Value response_json;
+    if (response.getStatus() == Poco::Net::HTTPResponse::HTTP_OK)
+    {
+        respStream >> response_json;
+    } else {
+        response_json = Json::Value::null;
+    }
+
+    //retrieve wanted element from result json, if not valid result return not_resolved.
+    std::string not_resolved = (on_not_resolvable == EMPTY) ? "" : name;
+    if (response_json.isNull())
+    {
+        return std::make_pair(name, not_resolved);
+    } else
+    {
+        Json::Value results = response_json["results"];
+
+        Json::Value val = results[0].get(key, not_resolved);
```

Comment:
what if there was no hit?

Reply:
It returns `not_resolved` if `results[0]`is out of bounds. But I'll make this more explicit by checking if results is empty.